### PR TITLE
feat(context-py): add session manager integration

### PR DESCRIPTION
**Pull request by lizradway**
## Description
Adds cross-session context management saturated by the session manager

## Type of Change

New feature

## Testing

How have you tested the change? Verify that the changes do not break functionality or introduce new warnings.

- [ ] I ran `hatch run prepare`

## Checklist
- [ ] I have read the CONTRIBUTING document
- [ ] I have reviewed and understand every line of code in this PR, including any generated by AI tools, and I can explain why it works
- [ ] My change is focused and reasonably small; I have split unrelated work into separate PRs
- [ ] I have added any necessary tests that prove my fix is effective or my feature works
- [ ] I have updated the documentation accordingly
- [ ] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: strands-py/src/strands/_context_manager/context_manager.py [hunk 1]
```diff
@@ -72,15 +72,30 @@ def __init__(
...
+
+        When True, stash data does not need to be embedded in session snapshots.
+        """
+        return self._stash_is_durable
+
     def init_agent(self, agent: Agent) -> None:
         """Register strategy hooks for proactive compression and overflow recovery."""
         if not self._stash_disabled:
             storage = self._stash_explicit_storage or getattr(agent, "storage", None) or InMemoryStorage()
+            self._stash_is_durable = not isinstance(storage, InMemoryStorage)
```
Comment (strandly-the-agent) [inline strands-py/src/strands/_context_manager/context_manager.py]: 🟡 **Should-fix — any ephemeral storage that isn't literally an `InMemoryStorage` instance is classified durable, so the snapshot only stores an external pointer and every offloaded block is unrecoverable after a restart, with no warning.**

`stash_is_durable` is `True` for `InMemoryStorage().namespace("tenant")`, for `_NamespacedStorage(InMemoryStorage(), …)`, and for any third-party in-memory backend; the docstring above promises "survives process restarts", which this line doesn't compute. Reached via `Agent(storage=InMemoryStorage().namespace("tenant"), context_manager=ContextManager(), session_manager=SnapshotSessionManager(...))` — dangling refs then surface to the model as `Error: reference not found`.

Suggestion: mirror the existing `_NAMESPACED` sentinel with an `_EPHEMERAL` one set on `InMemoryStorage` and copied by `_NamespacedStorage` / `namespace()` — the same shape TS uses (`storage.ts:18`, `:204-205`). Cost: ~8 lines across three files, and it touches `storage/`; custom ephemeral backends have to opt in. Fix this together with the `Stash.__init__` prefix comment below — fixing this alone makes that one worse.
```suggestion
            self._stash_is_durable = getattr(storage, "_ephemeral", None) is not _EPHEMERAL
```

File: strands-py/src/strands/agent/agent.py [hunk 2]
```diff
@@ -404,6 +404,12 @@ def __init__(
             context_manager, conversation_manager, plugins
         )
 
+        from .._context_manager.context_manager import ContextManager as _ContextManager
+
+        self._context_manager: ContextManager | None = (
+            context_manager if isinstance(context_manager, _ContextManager) else None
+        )
```
Comment (strandly-the-agent) [inline strands-py/src/strands/agent/agent.py]: 🟡 **Should-fix — `Agent(plugins=[ContextManager()])` gives a fully working stash that is never persisted: `agent.context_manager` is `None`, so no `stash` key is ever written and offloaded content is lost on restore, silently.**

Verified: plugin registered, live stash entries present, `"stash" not in snapshot.data`. The property docstring ("if one is registered on this agent") is also inaccurate for this path. `ContextManager` is a `Plugin`, so the plugin list is a natural place to put it.

Suggestion: refuse the ambiguity the way `memory_manager` does a few lines below (`agent.py:596-601`):
```python
if "strands:context-manager" in self._plugin_registry._plugins and self._context_manager is None:
    raise ValueError(
        "A ContextManager is already registered via plugins; pass it through the "
        "context_manager parameter instead"
    )
```
Cost: ~4 lines, no TS drift (TS shares the gap). Cheaper alternative: document in the docstring that the property is `None` for `"auto"`/`"agentic"` and for plugin-list registration.

File: strands-py/src/strands/_context_manager/stash.py [hunk 3]
```diff
@@ -58,8 +58,15 @@ class Stash:
     """Namespaced storage wrapper for persisting offloaded content blocks."""
 
     def __init__(self, storage: Storage, session_id: str, agent_id: str) -> None:
+        self._base_storage = storage
+        self._session_id = session_id
         self._storage = _resolve_namespace(storage, f"{STASH_PREFIX}/{session_id}/scopes/agent/{agent_id}")
```
Comment (strandly-the-agent) [inline strands-py/src/strands/_context_manager/stash.py]: 🟡 **Should-fix — when the storage handed to `Stash` is already namespaced, the stash prefix collapses: keys are written with no `context/<session>/…` prefix, so `clear_session()` matches nothing and `delete_session()` leaves the stash behind — contradicting its new docstring "Delete all snapshots and stash data".**

`_resolve_namespace` returns the storage unchanged when it carries `_NAMESPACED`, so `self._storage is self._base_storage`. Reached via `Agent(storage=LocalFileStorage(d).namespace("app"), context_manager=ContextManager(), session_manager=SnapshotSessionManager("s1", storage=scoped))` — repro: two stash keys survive `delete_session()`.

Not a blocker on this head: dropping `clear()` in 609dfaf removed the destructive half (an unrelated `memory/user-notes.json` in the same view now survives). But if the durability check above becomes marker-based without this fix, `take_snapshot()` on a collapsed ephemeral view will inline every key in the view — session snapshots included.

Suggestion: always apply the prefix, as TS `stash.ts:51` does. Cost: `_NamespacedStorage` import; sessions written before the fix leave orphaned unprefixed keys (no in-repo caller does this today).
```suggestion
        self._storage = _NamespacedStorage(storage, f"{STASH_PREFIX}/{session_id}/scopes/agent/{agent_id}")
```

File: strands-py/src/strands/session/snapshot_session_manager.py [hunk 4]
```diff
@@ -523,3 +535,87 @@ def _capture(self, agent: "Agent") -> Snapshot:
...
+            if entries:
+                snapshot.data["stash"] = {
+                    "location": "inline",
+                    "entries": entries,
+                }
+        except Exception:
+            logger.warning(
+                "session_id=<%s> | failed to include stash data in snapshot, saving without stash",
+                self.session_id,
+            )
```
Comment (strandly-the-agent) [inline strands-py/src/strands/session/snapshot_session_manager.py]: 🟡 **Should-fix (introduced by the hardening commit) — one failed stash read during save overwrites `snapshot_latest` with a snapshot that has no `stash` key, erasing already-persisted inline entries while the messages keep their `[ref: …]` placeholders.**

Confirmed: baseline snapshot carried 2 inline refs → one flaky `read` → persisted snapshot has no stash key → after a simulated restart `retrieve(ref)` is `None` and the placeholder survives. A transient *session* write failure in the same `sync_agent` still propagates, so stash persistence is now strictly weaker than message persistence, and `logger.warning` drops the traceback. Narrow today (the inline branch only runs for `InMemoryStorage`, whose `read` doesn't raise) — it becomes real as soon as durability is marker-based and custom ephemeral backends take this path.

Suggestion: keep the swallow on restore/delete, fail loudly on save (or carry the previous snapshot's stash payload forward). Cost: `test_save_succeeds_when_stash_storage_fails` currently asserts the erasing behaviour and would need updating. Minimum: `logger.exception` so the cause survives.
```suggestion
        except Exception:
            logger.exception("session_id=<%s> | failed to include stash data in snapshot", self.session_id)
            raise
```

File: strands-py/src/strands/session/snapshot_session_manager.py [hunk 5]
```diff
@@ -523,3 +535,87 @@ def _capture(self, agent: "Agent") -> Snapshot:
...
+
+        try:
+            if context_manager.stash_is_durable:
+                snapshot.data["stash"] = {
+                    "location": "external",
+                    "storage_type": context_manager.stash.storage_type_name,
+                }
+                return
+
+            entries = await context_manager.stash.take_snapshot()
```
Comment (strandly-the-agent) [inline strands-py/src/strands/session/snapshot_session_manager.py]: 🟡 **Design question — in the default configuration every snapshot write embeds a full copy of L1: with 20 KB entries, `snapshot_latest` measured 20 / 100 / 200 KB after 1 / 5 / 10 offloads, rewritten on every save, so cumulative write volume grows quadratically.**

The default pairing splits storage: the stash falls back to a fresh `InMemoryStorage` (inline branch) while the session manager defaults to `LocalFileStorage`. Reached by `Agent(context_manager=ContextManager(), session_manager=SnapshotSessionManager(...))` as soon as anything is offloaded; per message rather than per invocation under `save_latest_on="message"`, on the synchronous invocation path.

Suggestion (a decision, not a patch): when a session manager is attached and no explicit stash storage was given, default the stash to the storage the session manager resolved so the external branch is taken. Cost: changes a default and is a cross-SDK call — TS behaves identically and the design doc defers eviction past v1 — so this may be a deliberate tradeoff. Would you rather take it now or note it as known?

File: strands-py/src/strands/session/snapshot_session_manager.py [hunk 6]
```diff
@@ -476,7 +486,9 @@ async def _save_immutable_and_latest(self, agent: "Agent") -> None:
         can leave an orphaned immutable snapshot (harmless — the next list simply includes it)
         but never a ``snapshot_latest`` pointing at history that was never written.
         """
-        data = _serialize_snapshot(self._capture(agent))
+        snapshot = self._capture(agent)
+        await self._include_stash_data(agent, snapshot)
```
Comment (strandly-the-agent) [inline strands-py/src/strands/session/snapshot_session_manager.py:493]: 🟡 **Should-fix (tests) — this call site is entirely uncovered: deleting it leaves all 319 tests green.**

A user with a `snapshot_trigger` plus an ephemeral stash would get immutable checkpoints with no stash at all, so `restore_snapshot(snapshot_id=…)` restores placeholders that can never resolve. No test in `TestSnapshotStashIntegration` constructs a manager with `snapshot_trigger`.

Suggestion: one test — manager with `snapshot_trigger=lambda **_: True` and an `InMemoryStorage` stash, `load_snapshot` a ref, run a second invocation, read the immutable snapshot key, assert `data["stash"]["location"] == "inline"`.

File: strands-py/src/strands/_context_manager/stash.py [hunk 7]
```diff
@@ -110,6 +117,47 @@ async def delete(self, reference: str) -> None:
...
+            await self._storage.delete(key)
+
+    async def clear_session(self) -> None:
+        """Delete all stash data for this session across all agents.
+
+        Unlike :meth:`clear`, which is scoped to this agent's namespace, this
+        scans ``context/<session_id>/`` on the base storage to remove data from
+        every agent that wrote to the session.
+        """
+        prefix = f"{STASH_PREFIX}/{self._session_id}/"
```
Comment (strandly-the-agent) [inline strands-py/src/strands/_context_manager/stash.py:156]: 🟡 **Should-fix (tests) — neither guarantee `clear_session` makes is pinned: dropping the trailing slash here survives the suite, and so does swapping `clear_session()` for the agent-scoped `clear()` in `_delete_stash_data`.**

The existing tests use `sess-1` / `sess-2`, which a broadened prefix still passes; prefix broadening is exactly the class of bug the repo already guards for `delete_session` (`test_snapshot_session_manager.py:166`).

Suggestion: two short tests — (a) boundary: sessions `s1` and `s10`, clear `s1`, assert `s10` still lists `["tool-1_0"]`; (b) in `test_delete_session_clears_stash`, add a second agent on the same session so `clear_session()` becomes load-bearing:
```python
other = Stash(stash_mem, "s1", "a2")
await other.load_snapshot({"ref-2": {"text": "other-agent"}})
await manager.delete_session()
assert await other.list() == []
```

**Review by strandly-the-agent — CHANGES_REQUESTED** [review]
**Changes requested** — faithful port of the TS stash↔snapshot integration and the happy paths work end to end (durable, inline, and ephemeral→durable migration all verified after a simulated restart). No blockers; 7 should-fix items, 3 of them one coupled storage-classification problem: the `isinstance(storage, InMemoryStorage)` durability check, the collapsed stash prefix on pre-namespaced storage, and the new swallow-on-save — each on its own makes offloaded content silently unrecoverable after restart. The `plugins=[ContextManager()]` path persists nothing at all.

**API review:** this adds a public `Agent.context_manager` property plus a persisted `stash` snapshot field and carries no `api/needs-review` label — needs the label and the *Public API Changes* section in the description (`team/PR.md`, `team/API_BAR_RAISING.md`). One API reviewer should suffice; not an API-meeting-sized change.

Reviewed head `609dfaf`. Details, questions, reading order and non-blocking items below.

<details><summary>Verified</summary>

- ✅ Branch `lizradway/python-context-session-manager-integration` @ `609dfaf08` (merge-base `545d9ed07`)
- ✅ `pytest tests/strands/session/test_snapshot_session_manager.py tests/strands/_context_manager tests/strands/agent/test_agent_context_manager.py tests/strands/agent/test_agent_storage.py` → **319 passed**
- ✅ `ruff check` on the touched files → 1 `I001` (see appendix); `ruff format --check` drift in `retrieval_tool.py` is pre-existing on main
- ✅ Repros run on `609dfaf`: durability misclassification of `InMemoryStorage().namespace(...)`; plugins-path `agent.context_manager is None` with a populated stash and no `stash` key in the snapshot; prefix collapse with `LocalFileStorage(d).namespace("app")` (stash keys survive `delete_session`; the destructive delete from the previous head is gone); inline snapshot growth 20→200 KB over 10 offloads; swallow-on-save erasing persisted inline entries
- ✅ Verified non-issues, so nobody re-litigates: `storage_type` vs TS `storageType` is fine (the payload is already language-cased and TS snapshots can't load in Python anyway); the one-message lag under `save_latest_on="message"` is benign (compression happens at `BeforeModelCall`, so the snapshot holds the uncompressed message); no path traversal via `load_snapshot` keys
- ⚠️ Process disclosure: the adversarial pass failed to return twice; its one repro (the swallow-on-save item) was run by the orchestrator and is behaviour-verified, but it's the only finding without an independent pass behind it

</details>

<details><summary>Per-pass breakdown</summary>

- **Correctness** — 5 findings, 5 appendix; verdict changes-requested. Durability check, prefix collapse, inline growth, plugins path, missing `stash=False` warning from design §3.1.
- **API bar-raiser** — verdict: needs `api/needs-review` label, not a meeting. Shapes win on TS parity; the one mechanism failure is the `isinstance` durability check. Raised the `stash_is_durable`/`storage_type_name` placement, `take_snapshot`/`load_snapshot` naming, and `Stash.clear()` having no caller after `609dfaf`.
- **Test quality** — 14 mutations, 9 survived. Untested: the immutable-checkpoint call site, `clear_session` prefix boundary, `clear()` vs `clear_session()`, the real `store_message` → snapshot → restore path (binary round-trip works today, nothing pins it), `restore_snapshot(snapshot_id)`.
- **Adversarial** — did not complete (see disclosure above).
- Aggregation: 29 raw findings → 7 inline, 8 questions, 11 appendix; 11 merged/suppressed (already fixed in `609dfaf`, bot re-nags, verified non-issues).

</details>

<details><summary>Questions</summary>

- **(blocking)** Does the `plugins=[ContextManager()]` path stay silent, or raise like `memory_manager`? What drove the revert from the registry lookup in `498587c5d` — TS parity, or a concrete problem with it?
- **(blocking)** Does the `_EPHEMERAL` marker land in this PR or a prerequisite PR touching `storage/`? It gates three of the inline items, and the durability fix without the prefix fix is a regression.
- Should the stash default to the session manager's resolved storage when one is attached, so the default config stops inlining L1 into every write?
- Design §3.1 (`team/designs/0015-context-manager.md:264`) asks for a warning when `stash=False` is combined with a session manager; `_include_stash_data` returns silently and offload is still destructive. Deferred deliberately, or an oversight? `initialize()` looks like the natural home.
- `ContextManager.stash_is_durable` and `Stash.storage_type_name` are two facts about the same storage on two objects — why not `Stash.is_durable`? (Cost: TS drift.)
- `Stash.take_snapshot() -> dict` / `load_snapshot(entries)` reuse the `Agent.take_snapshot() -> Snapshot` verb pair with a different shape, and `load_snapshot` merges rather than replaces (undocumented). Worth `export_entries`/`import_entries` in both SDKs, or is parity the tiebreak?
- `restore_snapshot(snapshot_id=…)` time travel leaves the live stash as the union of current + snapshot entries. Intended? Either answer deserves a test.
- Is `Stash.clear()` intended public surface? After `609dfaf` nothing in `src/strands` calls it, and its session-wide sibling arguably belongs on the session manager.

</details>

<details><summary>Reading order</summary>

1. `strands-py/src/strands/session/snapshot_session_manager.py` — start at the `# -- Stash integration --` block (`_include_stash_data` / `_restore_stash_data` / `_delete_stash_data`), then the four call sites (`initialize`, `save_snapshot`, `_save_immutable_and_latest`, `_restore`, `delete_session`).
2. `strands-py/src/strands/_context_manager/context_manager.py` — the `stash` / `stash_is_durable` properties and the one-line durability decision in `init_agent`.
3. `strands-py/src/strands/_context_manager/stash.py` — `take_snapshot` / `load_snapshot` / `clear` / `clear_session`, with `__init__`'s `_resolve_namespace` in mind.
4. `strands-py/src/strands/agent/agent.py` — the `_context_manager` capture and property; compare with `_resolve_context_manager` and the `memory_manager` guard just below.
5. Tests, `TestSnapshotStashIntegration` first.

</details>

<details><summary>Appendix — non-blocking (11)</summary>

1. ⚪ `snapshot_session_manager.py:310-312` — `_agent_stash` is never reset when `context_manager is None`, so a reused manager keeps the previous agent's stash; TS assigns unconditionally. `self._agent_stash = context_manager.stash if context_manager is not None else None`.
2. ⚪ `snapshot_session_manager.py:431` — "Delete all snapshots and stash data" only holds if *this* manager instance was initialized by a stash-bearing agent; a cleanup-only manager leaves `context/<session>/` behind. Narrow the docstring or document the precondition.
3. ⚪ `agent.py:409-411` duplicates the `isinstance` already in `_resolve_context_manager` with a second function-local import — have the resolver return the resolved manager.
4. ⚪ `stash.py:66-68` — `storage_type_name` reports `_NamespacedStorage` (a private class name, now persisted) for namespaced views, and two `S3Storage` instances on different buckets compare equal, so the realistic "I moved my storage" case never warns. Unwrap `_NamespacedStorage._storage` when naming.
5. ⚪ `stash.py:134-141` — `load_snapshot` merges into the live stash rather than replacing it; docstring doesn't say so.
6. ⚪ Remaining test gaps: durable stash end-to-end through a restart and through `delete_session`; `restore_snapshot(snapshot_id)` with a stash; `_delete_stash_data` when `_agent_stash is None`; inline entries with real binary content; the negative case of the storage-type mismatch warning; unknown `location` values; `stash.py:130`'s `if data is not None`; the plugins path.
7. ⚪ `tests/strands/_context_manager/test_stash.py:6-7` — ruff `I001` (blank line between `import pytest` and the `strands` imports was removed); `ruff check --fix` clears it.
8. pre-existing — `ruff format --check` flags `src/strands/_context_manager/retrieval_tool.py` on main; not this PR.
9. pre-existing — Python's `LocalAgent` protocol declares neither `storage` nor `context_manager`, while TS `types/agent.ts` declares `contextManager`.
10. pre-existing — a duck-typed agent (like the bare `MagicMock` in `test_agent_storage.py`) makes `_agent_stash` a mock; a save would put a `MagicMock` in `storage_type` and fail `json.dumps`. No caller reaches it.
11. ✅ `storage_type` vs TS `storageType` — not a wire-compat bug; no change needed.

</details>

Evidence (test log, repro scripts + outputs, context pack, raw findings) is uploaded under `strands-agents/harness-sdk/pr/4254/` in the artifacts bucket.

**Review by strandly-the-agent — APPROVED** [review]
**Approve** — all six should-fix items from my last review are fixed in `87369cf` and I re-verified each against the code, not just the tests (repro output below). No new issues found; the fixes hold under the edge cases I tried (nested `namespace()` views, same instance via both `context_manager=` and `plugins=`, failing stash read on both `sync_agent` and the end-of-invocation save).

Still outstanding, neither blocking from my side: the `api/needs-review` label + *Public API Changes* section in the description (`Agent.context_manager` is public surface), and the open design question on the inline-L1 default (thread above) — happy for that to be a "known, deferred" answer.

<details><summary>Status of previous findings</summary>

| # | Finding | Status |
|---|---|---|
| 1 | `isinstance(InMemoryStorage)` durability check | ✅ fixed — `_EPHEMERAL` sentinel on `InMemoryStorage`, propagated by `_NamespacedStorage`; verified for `.namespace("x")`, nested `.namespace("x").namespace("y")`, direct `_NamespacedStorage`, and that `LocalFileStorage(...).namespace()` stays durable |
| 2 | `plugins=[ContextManager()]` silently unpersisted | ✅ fixed — raises `ValueError` pointing at `context_manager=`; also raises for `context_manager="auto"` + plugin (reasonable). Same instance via both paths hits the pre-existing "plugin already registered" error, fine |
| 3 | Stash prefix collapse on pre-namespaced storage | ✅ fixed — always `_NamespacedStorage`; on `LocalFileStorage(d).namespace("app")` keys now land under `context/s1/scopes/agent/a1/`, `delete_session()` removes them and leaves unrelated keys alone |
| 4 | Swallow-on-save erasing persisted inline entries | ✅ fixed — `_include_stash_data` propagates; a failing read during `sync_agent` raises and `snapshot_latest` keeps its previous stash payload. Restore/delete still swallow, as intended |
| 5 | Inline L1 in every write (default config) | ❓ open question, non-blocking |
| 6 | Immutable-checkpoint call site untested | ✅ fixed — `test_immutable_snapshot_includes_stash` |
| 7 | `clear_session` prefix boundary / cross-agent delete | ✅ (a) `test_prefix_boundary_does_not_broaden` added; (b) the cross-agent assertion in `test_delete_session_clears_stash` wasn't — `clear_session()` vs `clear()` is still indistinguishable by the suite. Nit, not re-opening |

</details>

<details><summary>Verified</summary>

- ✅ head `87369cf073ff8903a3d8d2620fe18947bbb6553a`; delta vs `609dfaf` is 10 files, +114/−43, all addressing review items — reviewed inline, no fan-out
- ✅ `pytest tests/strands/session/test_snapshot_session_manager.py tests/strands/_context_manager tests/strands/agent/test_agent_context_manager.py tests/strands/agent/test_agent_storage.py tests/strands/storage` → **446 passed**
- ✅ `ruff check` clean on touched files (the `I001` is gone)
- ✅ fix-verification script run on `87369cf`: 12/12 PASS (uploaded as `repro-verify-fixes-87369cf.{py,out}` in the artifacts bucket)

</details>

<details><summary>Nits (3, non-blocking)</summary>

- `snapshot_session_manager.py` — the deleted `test_save_succeeds_when_stash_storage_fails` wasn't replaced with its inverse; a 5-line test asserting `sync_agent` raises and `snapshot_latest` is unchanged would pin the new contract (it's what my repro checks).
- `context_manager.py:98` — `from ..storage.storage import _EPHEMERAL` is a function-local import inside `init_agent`; `..storage.in_memory_storage` is already imported at module level and itself imports `_EPHEMERAL`, so there's no cycle — it can move to the top.
- `test_snapshot_session_manager.py:980-982` — `ruff format` (0.16.x, within the repo's `<0.17` pin) wants the `SnapshotSessionManager(...)` call on one line.

</details>

File: strands-py/src/strands/_context_manager/context_manager.py (hunk 1 above)
Comment (strandly-the-agent) [reply strands-py/src/strands/_context_manager/context_manager.py]: Verified on `87369cf`: `_EPHEMERAL` propagates through single and nested `namespace()` views; `LocalFileStorage.namespace()` stays durable. Fixed.

File: strands-py/src/strands/agent/agent.py (hunk 2 above)
Comment (strandly-the-agent) [reply strands-py/src/strands/agent/agent.py]: Verified on `87369cf`: `Agent(plugins=[ContextManager()])` now raises with a pointer to `context_manager=`. Fixed.

File: strands-py/src/strands/_context_manager/stash.py (hunk 3 above)
Comment (strandly-the-agent) [reply strands-py/src/strands/_context_manager/stash.py]: Verified on `87369cf`: keys land under `context/<session>/scopes/agent/<id>/` inside a pre-namespaced view and `delete_session()` removes them. Fixed.

File: strands-py/src/strands/session/snapshot_session_manager.py (hunk 4 above)
Comment (strandly-the-agent) [reply strands-py/src/strands/session/snapshot_session_manager.py]: Verified on `87369cf`: a failing stash read during `sync_agent` propagates and `snapshot_latest` keeps its previous stash payload. Fixed.

File: strands-py/src/strands/session/snapshot_session_manager.py (hunk 6 above)
Comment (strandly-the-agent) [reply strands-py/src/strands/session/snapshot_session_manager.py:493]: Verified on `87369cf`: `test_immutable_snapshot_includes_stash` covers the `_save_immutable_and_latest` call site. Fixed.

File: strands-py/src/strands/_context_manager/stash.py (hunk 7 above)
Comment (strandly-the-agent) [reply strands-py/src/strands/_context_manager/stash.py:156]: Verified on `87369cf`: prefix-boundary test added. The cross-agent `delete_session` assertion (b) wasn't, so `clear_session()` vs `clear()` is still undistinguished by the suite — nit only, not re-opening.

File: strands-py/src/strands/session/snapshot_session_manager.py [hunk 8]
```diff
@@ -523,3 +535,83 @@ def _capture(self, agent: "Agent") -> Snapshot:
...
+                "session_id=<%s> | failed to restore stash data from snapshot, continuing without stash",
+                self.session_id,
+            )
+
+    async def _delete_stash_data(self) -> None:
+        """Delete all stash data during session deletion.
+
+        Storage errors are logged and swallowed so a stash failure never prevents session deletion.
+        """
+        if self._agent_stash is not None:
```
Comment (opieter-aws) [inline strands-py/src/strands/session/snapshot_session_manager.py]: `_agent_stash` is set only in `initialize()`. The docs builds a bare `SnapshotSessionManager(...)` and calls `delete_session()`, which doesn't work (it never initialized, so stash is orphaned while snapshots are deleted). Either when `_agent_stash is None`, fall back to deleting `context/<session_id>/` on the base storage with a warning, or soften the docs.

File: strands-py/src/strands/_context_manager/context_manager.py [hunk 9]
```diff
@@ -72,15 +72,32 @@ def __init__(
...
+
+        When True, stash data does not need to be embedded in session snapshots.
+        """
+        return self._stash_is_durable
+
     def init_agent(self, agent: Agent) -> None:
         """Register strategy hooks for proactive compression and overflow recovery."""
         if not self._stash_disabled:
             storage = self._stash_explicit_storage or getattr(agent, "storage", None) or InMemoryStorage()
+            from ..storage.storage import _EPHEMERAL
```
Comment (opieter-aws) [inline strands-py/src/strands/_context_manager/context_manager.py]: Intentional mid-file import?

Comment (lizradway) [reply strands-py/src/strands/_context_manager/context_manager.py]: nope can move!

**Review by strandly-the-agent — APPROVED** [review]
**Approve** on `aa235ef` — the delta since my last approval (`87369cf`) is the uninitialized-`delete_session()` stash fallback `opieter-aws` asked for, the `_EPHEMERAL` import moved to module level, and the `clear()` parity call restored. I attacked the fallback and it holds: it respects the `context/<session>/` prefix boundary (`s1` vs `s10`), works on a pre-namespaced shared view, leaves unrelated keys and other sessions' stash alone, and a CM-less agent's `delete_session()` doesn't sweep other sessions. Nothing new to raise.

Still open, non-blocking: `api/needs-review` label + *Public API Changes* in the description (bot has this covered), and my inline-L1-default design question — fine as "known / deferred".

<details><summary>Verified</summary>

- ✅ head `aa235efedc8880de0ed440ecd762790d63a289ab`; delta `87369cf..aa235ef` = 3 files, +49/−10, reviewed inline
- ✅ `pytest tests/strands/session/test_snapshot_session_manager.py tests/strands/_context_manager tests/strands/agent/test_agent_context_manager.py tests/strands/agent/test_agent_storage.py tests/strands/storage` → **447 passed**; `ruff check` clean
- ✅ fallback repro 6/6 PASS (`repro-verify-fallback-aa235ef.{py,out}` in the artifacts bucket); bare manager with no storage still raises the pre-existing `RuntimeError` rather than silently skipping
- The redundant `clear()` before `clear_session()` is a deliberate TS-parity choice already settled in the thread above; not re-raising

</details>

**Review by opieter-aws — APPROVED** [review]
